<a href="https://colab.research.google.com/github/iam-SarthakVerma/VehicleIQ/blob/main/Motor_Vehicles_registrations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Neural City GenAI Assignment
#### By Sarthak Verma



## Natural Language Analytics over Government Vehicle Registration Data

Dataset:
State/UT-wise Registered Motor Vehicles in India (2006–2016)

#### Drive link of folder which contains "screenshots of working model", "dataset.csv", downloaded version of this model, readme . " https://drive.google.com/drive/folders/1gQGxzPpe9j6i2BXEZPLUbuz1n744qCUI?usp=sharing " for your reference.

## 1. Load Dataset
Importing the government dataset into Pandas.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import pandas as pd

df = pd.read_csv(
    "/content/drive/MyDrive/NeuralCity/Dataset.csv"
)

df.head()

Mounted at /content/drive


,Category,States / UT-wise,Year - 2006,Year - 2007,Year - 2008,Year - 2009,Year - 2010,Year - 2011,Year - 2012,Year - 2013,...,Year - 2016,Note of Year - 2006,Note of Year - 2007,Note of Year - 2008,Note of Year - 2010,Note of Year - 2011,Note of Year - 2013,Note of Year - 2014,Note of Year - 2015,Note of Year - 2016
0,State,Andhra Pradesh,7218.195,6366.856,7207.946,8058.948,8923.139,10189.347,12424.0,6286.419,...,8727.558,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,State,Arunachal Pradesh,22.101,22.101,22.101,22.101,22.101,144.534,151.0,151.279,...,265.000,*,*,*,*,NaN,#,#,#,NaN
2,State,Assam,914.023,1021.036,1116.422,1234.751,1383.721,1582.128,1807.0,1877.522,...,2817.390,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,State,Bihar,1432.343,1577.383,1739.140,1959.553,2356.986,2673.209,3113.0,3617.227,...,5482.329,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,State,Chhattisgarh,1540.849,1733.809,1934.871,2114.589,2435.773,2766.037,3104.0,3437.242,...,4810.452,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
import os

print(os.listdir("/content/drive/MyDrive/NeuralCity"))

['Dataset.csv', "Sarthak_Verma's_Assignment", 'Screenshots of working model']


In [ ]:
import pandas as pd

df = pd.read_csv(
    "/content/drive/MyDrive/NeuralCity/Dataset.csv"
)

df.head()

,Category,States / UT-wise,Year - 2006,Year - 2007,Year - 2008,Year - 2009,Year - 2010,Year - 2011,Year - 2012,Year - 2013,...,Year - 2016,Note of Year - 2006,Note of Year - 2007,Note of Year - 2008,Note of Year - 2010,Note of Year - 2011,Note of Year - 2013,Note of Year - 2014,Note of Year - 2015,Note of Year - 2016
0,State,Andhra Pradesh,7218.195,6366.856,7207.946,8058.948,8923.139,10189.347,12424.0,6286.419,...,8727.558,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,State,Arunachal Pradesh,22.101,22.101,22.101,22.101,22.101,144.534,151.0,151.279,...,265.000,*,*,*,*,NaN,#,#,#,NaN
2,State,Assam,914.023,1021.036,1116.422,1234.751,1383.721,1582.128,1807.0,1877.522,...,2817.390,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,State,Bihar,1432.343,1577.383,1739.140,1959.553,2356.986,2673.209,3113.0,3617.227,...,5482.329,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,State,Chhattisgarh,1540.849,1733.809,1934.871,2114.589,2435.773,2766.037,3104.0,3437.242,...,4810.452,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## 2. Dataset Size Inspection

Checking the number of rows and columns in the dataset.

In [ ]:
print(df.shape)

(50, 22)


## 3. Dataset Schema Inspection

Examining available columns and attributes.

In [ ]:
print(df.columns.tolist())

['Category', 'States / UT-wise', 'Year - 2006', 'Year - 2007', 'Year - 2008', 'Year - 2009', 'Year - 2010', 'Year - 2011', 'Year - 2012', 'Year - 2013', 'Year - 2014', 'Year - 2015', 'Year - 2016', 'Note of Year - 2006', 'Note of Year - 2007', 'Note of Year - 2008', 'Note of Year - 2010', 'Note of Year - 2011', 'Note of Year - 2013', 'Note of Year - 2014', 'Note of Year - 2015', 'Note of Year - 2016']


## 4. Inspect Final Records

Reviewing the last rows to identify metadata and summary records.

In [ ]:
df.tail()

,Category,States / UT-wise,Year - 2006,Year - 2007,Year - 2008,Year - 2009,Year - 2010,Year - 2011,Year - 2012,Year - 2013,...,Year - 2016,Note of Year - 2006,Note of Year - 2007,Note of Year - 2008,Note of Year - 2010,Note of Year - 2011,Note of Year - 2013,Note of Year - 2014,Note of Year - 2015,Note of Year - 2016
45,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
46,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
47,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
48,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
49,NaN,37,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## 5. Data Type Inspection

Understanding column types and missing values.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Category             39 non-null     object 
 1   States / UT-wise     40 non-null     object 
 2   Year - 2006          38 non-null     float64
 3   Year - 2007          38 non-null     float64
 4   Year - 2008          38 non-null     float64
 5   Year - 2009          38 non-null     float64
 6   Year - 2010          38 non-null     float64
 7   Year - 2011          38 non-null     float64
 8   Year - 2012          38 non-null     float64
 9   Year - 2013          39 non-null     float64
 10  Year - 2014          39 non-null     float64
 11  Year - 2015          39 non-null     float64
 12  Year - 2016          39 non-null     float64
 13  Note of Year - 2006  1 non-null      object 
 14  Note of Year - 2007  1 non-null      object 
 15  Note of Year - 2008  1 non-null      objec

## 6. Initial Data Cleaning

Removing empty rows and records with missing state information.

In [ ]:
# Removing completely empty rows
df = df.dropna(how='all')

# Removing rows where state name is missing
df = df[df['States / UT-wise'].notna()]

print("Shape:", df.shape)

df.tail()

Shape: (40, 22)


,Category,States / UT-wise,Year - 2006,Year - 2007,Year - 2008,Year - 2009,Year - 2010,Year - 2011,Year - 2012,Year - 2013,...,Year - 2016,Note of Year - 2006,Note of Year - 2007,Note of Year - 2008,Note of Year - 2010,Note of Year - 2011,Note of Year - 2013,Note of Year - 2014,Note of Year - 2015,Note of Year - 2016
35,UT,Lakshadweep,6.263,6.740,7.235,7.235,7.855,8.753,10.0,11.105,...,16.008,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
36,UT,Puducherry,384.250,432.081,484.008,538.248,599.129,672.803,755.0,699.614,...,863.313,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
37,Total (UTs),Total (UTs),5665.127,6770.818,7281.199,7787.743,8506.144,9140.164,9420.0,9872.784,...,11742.977,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
38,Grand Total,Grand Total,89618.267,96707.260,105352.854,114951.033,127745.972,141865.607,159491.0,176043.945,...,230030.845,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
49,NaN,37,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## 7. Category Distribution Analysis

Inspecting the distribution of States, Union Territories and aggregate records.

In [ ]:
print(df['Category'].value_counts())

Category
State             29
UT                 7
Total (States)     1
Total (UTs)        1
Grand Total        1
Name: count, dtype: int64


## 8. Remove Metadata Columns

Dropping note columns that are not required for analytical queries.

In [ ]:
cols_to_drop = [col for col in df.columns if "Note" in col]

df_clean = df.drop(columns=cols_to_drop)

print(df_clean.columns.tolist())

['Category', 'States / UT-wise', 'Year - 2006', 'Year - 2007', 'Year - 2008', 'Year - 2009', 'Year - 2010', 'Year - 2011', 'Year - 2012', 'Year - 2013', 'Year - 2014', 'Year - 2015', 'Year - 2016']


## 9. Transform Dataset into Long Format

Converting yearly columns into a normalized structure containing State, Year and Vehicle Count.

In [ ]:
year_cols = [col for col in df_clean.columns if "Year -" in col]

df_long = pd.melt(
    df_clean,
    id_vars=["Category", "States / UT-wise"],
    value_vars=year_cols,
    var_name="Year",
    value_name="Vehicles"
)

df_long.head()

,Category,States / UT-wise,Year,Vehicles
0,State,Andhra Pradesh,Year - 2006,7218.195
1,State,Arunachal Pradesh,Year - 2006,22.101
2,State,Assam,Year - 2006,914.023
3,State,Bihar,Year - 2006,1432.343
4,State,Chhattisgarh,Year - 2006,1540.849


## 10. Data Type Conversion

Converting year and vehicle count fields into numerical formats suitable for computation.

In [ ]:
df_long["Year"] = (
    df_long["Year"]
    .str.replace("Year - ", "", regex=False)
    .astype(int)
)

df_long["Vehicles"] = pd.to_numeric(
    df_long["Vehicles"],
    errors="coerce"
)

df_long.head()

,Category,States / UT-wise,Year,Vehicles
0,State,Andhra Pradesh,2006,7218.195
1,State,Arunachal Pradesh,2006,22.101
2,State,Assam,2006,914.023
3,State,Bihar,2006,1432.343
4,State,Chhattisgarh,2006,1540.849


## 11. Dataset Validation

Verifying that the transformed dataset has the expected structure and data types.

In [ ]:
print(df_long.shape)

print(df_long.info())

df_long.head()

(440, 4)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440 entries, 0 to 439
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Category          429 non-null    object 
 1   States / UT-wise  440 non-null    object 
 2   Year              440 non-null    int64  
 3   Vehicles          422 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 13.9+ KB
None


,Category,States / UT-wise,Year,Vehicles
0,State,Andhra Pradesh,2006,7218.195
1,State,Arunachal Pradesh,2006,22.101
2,State,Assam,2006,914.023
3,State,Bihar,2006,1432.343
4,State,Chhattisgarh,2006,1540.849


## 12. Standardize Column Names

Renaming columns to simplify downstream querying.

In [ ]:
df_long = df_long.rename(
    columns={
        "States / UT-wise": "State"
    }
)

df_long.head()

,Category,State,Year,Vehicles
0,State,Andhra Pradesh,2006,7218.195
1,State,Arunachal Pradesh,2006,22.101
2,State,Assam,2006,914.023
3,State,Bihar,2006,1432.343
4,State,Chhattisgarh,2006,1540.849


## 13. Sample Analytical Query

Inspecting the highest vehicle registrations in 2016.

In [ ]:
df_long[df_long["Year"] == 2016]\
    .sort_values(
        "Vehicles",
        ascending=False
    )\
    .head(10)

,Category,State,Year,Vehicles
438,Grand Total,Grand Total,2016,230030.845
429,Total (States),Total (States),2016,218287.868
414,State,Maharashtra,2016,27869.866
423,State,Tamil Nadu,2016,24203.348
427,State,Uttar Pradesh,2016,23936.366
406,State,Gujarat,2016,20361.296
411,State,Karnataka,2016,16291.765
421,State,Rajasthan,2016,13632.086
437,Total (UTs),Total (UTs),2016,11742.977
413,State,Madhya Pradesh,2016,11141.041


## 14. Recheck Category Distribution

Confirming category counts after transformation.

In [ ]:
print(df['Category'].value_counts())

Category
State             29
UT                 7
Total (States)     1
Total (UTs)        1
Grand Total        1
Name: count, dtype: int64


## 15. Verify Transformed Dataset Size

Checking the final dataset dimensions.

In [ ]:
print(df_long.shape)

(440, 4)


## 16. Remove Aggregate Records

Excluding summary rows to ensure analysis is performed only on individual States and Union Territories.

In [ ]:
df_long = df_long[
    ~df_long["Category"].isin(
        [
            "Total (States)",
            "Total (UTs)",
            "Grand Total"
        ]
    )
]

print(df_long["Category"].value_counts())
print(df_long.shape)

Category
State    319
UT        77
Name: count, dtype: int64
(407, 4)


## 17. Gemini Configuration

Configuring Gemini for natural language query understanding and intent extraction.

In [ ]:
import google.generativeai as genai
from google.colab import userdata

genai.configure(
    api_key=userdata.get("GEMINI_API_KEY")
)

model = genai.GenerativeModel(
    "models/gemini-2.5-flash"
)

response = model.generate_content(
    "Say SUCCESS"
)

print(response.text)

/usr/local/lib/python3.12/dist-packages/google/colab/_import_hooks/_hook_injector.py:55: FutureWarning: 

All support for the `google.generativeai` package has ended. It will no longer be receiving 
updates or bug fixes. Please switch to the `google.genai` package as soon as possible.
See README for more details:

https://github.com/google-gemini/deprecated-generative-ai-python/blob/main/README.md

  loader.exec_module(module)


SUCCESS


#18 Groq Configuration

In [ ]:
!pip install groq

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.7/143.7 kB 3.0 MB/s eta 0:00:00


In [ ]:
from groq import Groq
from google.colab import userdata

groq_client = Groq(
    api_key=userdata.get("Groq")
)

response = groq_client.chat.completions.create(
    model="llama-3.3-70b-versatile",
    messages=[
        {
            "role": "user",
            "content": "Reply with only the word WORKING"
        }
    ]
)

print(response.choices[0].message.content)

WORKING


## 19. Query Planning Layer Using Gemini

Converting natural language questions into structured execution plans.

In [ ]:
import json

def get_query_plan_gemini(question):

    prompt = f"""
You are a data query planner.

Dataset Schema:

State : Name of State or UT
Year : 2006-2016
Vehicles : Registered Motor Vehicles

Your job:

1. Determine whether the question can be answered from this dataset.
2. If answerable, generate a JSON plan.
3. If not answerable, return out_of_scope.

Allowed operations:

rank
trend
compare
aggregate
growth

Examples:

Question:
Which state had the highest vehicles in 2016?

Output:
{{
"answerable": true,
"operation": "rank",
"year": 2016,
"order": "desc",
"limit": 1
}}

Question:
Compare Maharashtra and Gujarat

Output:
{{
"answerable": true,
"operation": "compare",
"states": ["Maharashtra","Gujarat"]
}}

Question:
Show trend of Kerala

Output:
{{
"answerable": true,
"operation": "trend",
"state": "Kerala"
}}

Question:
What was Delhi AQI in 2016?

Output:
{{
"answerable": false,
"reason": "Dataset contains vehicle registration data only."
}}

Question:
{question}

Return ONLY JSON.
"""

    print("Calling Gemini...")

    response = model.generate_content(prompt)

    print("Gemini responded")

    text = response.text.strip()

    print("Response text:")
    print(text)

    if text.startswith("```"):
        text = text.replace("```json", "")
        text = text.replace("```", "")
        text = text.strip()

    return json.loads(text)

#20. Groq Query Planner

In [ ]:
import json

def get_query_plan_groq(question):

    prompt = f"""
You are a query planner for a vehicle registration dataset.

Dataset:
- State
- Year (2006-2016)
- Vehicles

Return ONLY valid JSON.

Examples:

Question:
Which state had the highest registered vehicles in 2016?

Output:
{{
    "answerable": true,
    "operation": "rank",
    "year": 2016,
    "order": "desc",
    "limit": 1
}}

Question:
Top 5 states by registered vehicles in 2016.

Output:
{{
    "answerable": true,
    "operation": "rank",
    "year": 2016,
    "order": "desc",
    "limit": 5
}}

Question:
Show trend of Bihar.

Output:
{{
    "answerable": true,
    "operation": "trend",
    "state": "Bihar"
}}

Question:
Compare Maharashtra and Gujarat.

Output:
{{
    "answerable": true,
    "operation": "compare",
    "states": ["Maharashtra", "Gujarat"]
}}

Question:
What was the total number of registered vehicles in India in 2012?

Output:
{{
    "answerable": true,
    "operation": "aggregate",
    "year": 2012
}}

Question:
What was Delhi AQI in 2016?

Output:
{{
    "answerable": false,
    "reason": "Dataset contains vehicle registration data only."
}}

Now answer this question:

{question}

Return ONLY JSON.
"""

    response = groq_client.chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[
            {
                "role": "user",
                "content": prompt
            }
        ]
    )

    text = response.choices[0].message.content.strip()

    if text.startswith("```"):
        text = text.replace("```json", "")
        text = text.replace("```", "")
        text = text.strip()

    return json.loads(text)

#21. Query Planner Wrapper (Fallback Mechanism)

In [ ]:
# Global flags
gemini_available = True
groq_available = True


def reset_llm_flags():

    global gemini_available
    global groq_available

    gemini_available = True
    groq_available = True

    print("✅ Gemini and Groq re-enabled")


def normalize_plan(plan):

    if plan is None:
        return None

    if plan.get("operation") == "compare":

        plan.setdefault("year", 2016)
        plan.setdefault("over_years", False)

    if plan.get("operation") == "compare_all":

        plan.setdefault("year", 2016)

    return plan


def get_query_plan(question):

    global gemini_available
    global groq_available

    # Gemini
    if gemini_available:

        try:

            print("🤖 Calling Gemini...")

            plan = get_query_plan_gemini(question)

            return normalize_plan(plan)

        except Exception as e:

            print("⚠️ Gemini failed:", e)

            print(
                "🔄 Gemini disabled. "
                "Switching to Groq..."
            )

            gemini_available = False

    # Groq
    if groq_available:

        try:

            print("🚀 Calling Groq...")

            plan = get_query_plan_groq(question)

            return normalize_plan(plan)

        except Exception as e:

            print("⚠️ Groq failed:", e)

            print(
                "🔄 Groq disabled. "
                "Switching to Offline Engine..."
            )

            groq_available = False

    # Offline Engine
    print("🔋 Using Offline Engine")

    plan = local_query_plan(question)

    return normalize_plan(plan)

In [ ]:
def normalize_plan(plan):

    if plan is None:
        return None

    if plan.get("operation") == "compare":

        plan.setdefault("year", 2016)
        plan.setdefault("over_years", False)

    if plan.get("operation") == "compare_all":

        plan.setdefault("year", 2016)

    return plan

# 22. Query Execution Engine

Executing analytical operations on the dataset using Pandas.

In [ ]:
def execute_query(plan):

    if not plan["answerable"]:
        return {
            "type": "out_of_scope",
            "answer": plan["reason"]
        }

    operation = plan["operation"]

    if operation == "rank":

        filtered = df_long[
            df_long["Year"] == plan["year"]
        ]

        if "states" in plan:

            filtered = filtered[
                filtered["State"].isin(
                    plan["states"]
                )
            ]

        result = (
            filtered
            .sort_values(
                "Vehicles",
                ascending=(plan["order"] == "asc")
            )
            .head(plan["limit"])
        )

        return {
            "type": "table",
            "data": result
        }

    elif operation == "trend":

        states = plan.get(
            "states",
            [plan.get("state")]
        )

        start_year = plan.get(
            "start_year",
            2006
        )

        end_year = plan.get(
            "end_year",
            2016
        )

        result = (
            df_long[
                (df_long["State"].isin(states))
                &
                (df_long["Year"] >= start_year)
                &
                (df_long["Year"] <= end_year)
            ]
            .sort_values(["State", "Year"])
        )

        return {
            "type": "trend",
            "data": result
        }


    elif operation == "compare":

        year = plan.get("year", 2016)

        if plan.get("over_years", False):

            result = (
                df_long[
                    (df_long["State"].isin(plan["states"]))
                    &
                    (df_long["Year"] <= year)
                ]
                .sort_values(
                    ["State", "Year"]
                )
            )

        else:

            result = (
                df_long[
                    (df_long["State"].isin(plan["states"]))
                    &
                    (df_long["Year"] == year)
                ]
                .sort_values(
                    "Vehicles",
                    ascending=False
                )
            )

        return {
            "type": "compare",
            "data": result,
            "year": year,
            "over_years": plan.get(
                "over_years",
                False
            )
        }


    elif operation == "compare_all":

        result = (
            df_long[
                df_long["Year"] == plan["year"]
            ]
            .sort_values(
                "Vehicles",
                ascending=False
            )
        )

        if plan.get("limit"):

            result = result.head(
                plan["limit"]
            )

            target_state = plan["state"]

            if (
                target_state
                not in result["State"].values
            ):

                target_row = (
                    df_long[
                        (df_long["State"] == target_state)
                        &
                        (df_long["Year"] == plan["year"])
                    ]
                )

                result = pd.concat(
                    [
                        result,
                        target_row
                    ]
                )

        return {
            "type": "compare_all",
            "data": result,
            "state": plan["state"],
            "year": plan["year"]
        }

    elif operation == "aggregate":

        total = (
            df_long[
                df_long["Year"] == plan["year"]
            ]["Vehicles"]
            .sum()
        )

        return {
            "type": "number",
            "value": total
        }


    elif operation == "state_year":

        result = (
            df_long[
                (df_long["State"] == plan["state"])
                &
                (df_long["Year"] == plan["year"])
            ]
        )

        return {
            "type": "state_year",
            "data": result
        }


    elif operation == "growth":

        growth_df = []

        for state in df_long["State"].unique():

            state_data = (
                df_long[
                    df_long["State"] == state
                ]
                .sort_values("Year")
            )

            start = state_data.iloc[0]["Vehicles"]
            end = state_data.iloc[-1]["Vehicles"]

            growth = end - start

            growth_df.append(
                [state, growth]
            )

        growth_df = pd.DataFrame(
            growth_df,
            columns=["State", "Growth"]
        )

        result = (
            growth_df
            .sort_values(
                "Growth",
                ascending=False
            )
            .head(10)
        )

        return {
            "type": "growth",
            "data": result
        }

## 23. Local Answer Generation

Generating responses directly from computed results without additional LLM calls.

In [ ]:
def generate_answer_local(question, result):



    if result["type"] == "out_of_scope":
        return result["answer"]

    # Ranking Results
    if result["type"] == "table":

        df = result["data"]

        if len(df) == 1:

            row = df.iloc[0]

            return (
                f"{row['State']} had "
                f"{row['Vehicles']:.2f} "
                f"registered vehicles."
            )

        lines = []

        for i, (_, row) in enumerate(df.iterrows(), start=1):

            lines.append(
                f"{i}. {row['State']} : {row['Vehicles']:.2f}"
            )

        return "\n".join(lines)

    # Aggregate Results
    if result["type"] == "number":

        return (
            f"Total registered vehicles: "
            f"{result['value']:.2f}"
        )

    # Trend Results
    if result["type"] == "trend":

        states = result["data"]["State"].unique()

        if len(states) == 1:

            state = states[0]

            state_data = result["data"]

            start_year = state_data.iloc[0]["Year"]
            end_year = state_data.iloc[-1]["Year"]

            start = state_data.iloc[0]["Vehicles"]
            end = state_data.iloc[-1]["Vehicles"]

            growth = end - start

            return (
                f"{state} increased from "
                f"{start:.2f} vehicles in {start_year} "
                f"to {end:.2f} vehicles in {end_year}. "
                f"Net growth: {growth:.2f} vehicles."
            )

        lines = []

        for state in states:

            state_data = result["data"][
                result["data"]["State"] == state
            ]

            start_year = state_data.iloc[0]["Year"]
            end_year = state_data.iloc[-1]["Year"]

            start = state_data.iloc[0]["Vehicles"]
            end = state_data.iloc[-1]["Vehicles"]

            lines.append(
                f"{state}: {start:.2f} ({start_year}) → "
                f"{end:.2f} ({end_year})"
            )
        return "\n".join(lines)

    # Compare Results
    if result["type"] == "compare":

        if result.get("over_years", False):

            states = result["data"]["State"].unique()

            lines = []

            for state in states:

                state_data = result["data"][
                    result["data"]["State"] == state
                ]

                start = state_data.iloc[0]["Vehicles"]
                end = state_data.iloc[-1]["Vehicles"]

                lines.append(
                    f"{state}: "
                    f"{start:.2f} ({state_data.iloc[0]['Year']}) → "
                    f"{end:.2f} ({state_data.iloc[-1]['Year']})"
                  )

            return (
                "Comparison over years:\n"
                +
                "\n".join(lines)
            )

        latest = result["data"]

        latest = latest.sort_values(
            "Vehicles",
            ascending=False
        )

        lines = []

        for _, row in latest.iterrows():

            lines.append(
                f"{row['State']} : "
                f"{row['Vehicles']:.2f}"
            )

        return (
            f"Vehicle registrations in {result['year']}:\n"
            + "\n".join(lines)
        )

    elif result["type"] == "compare_all":

        state = result["state"]

        full_df = (
            df_long[
                df_long["Year"] == result["year"]
            ]
            .sort_values(
                "Vehicles",
                ascending=False
            )
            .reset_index(drop=True)
        )

        if state in full_df["State"].values:

            rank = (
                full_df[
                    full_df["State"] == state
                ]
                .index[0]
                + 1
            )

            vehicles = (
                full_df[
                    full_df["State"] == state
                ]["Vehicles"]
                .iloc[0]
            )

            return (
                f"In {result['year']}, "
                f"{state} ranked #{rank} "
                f"among all states "
                f"with {vehicles:.2f} "
                f"registered vehicles."
            )


    # State-Year Lookup Results
    if result["type"] == "state_year":

        if len(result["data"]) == 0:
            return "No data found."

        row = result["data"].iloc[0]

        return (
            f"{row['State']} had "
            f"{row['Vehicles']:.2f} "
            f"registered vehicles in "
            f"{int(row['Year'])}."
        )

    # Growth Results
    if result["type"] == "growth":

        top = result["data"].iloc[0]

        return (
            f"{top['State']} experienced the "
            f"highest growth of "
            f"{top['Growth']:.2f} vehicles "
            f"between 2006 and 2016."
        )

    return "Computation completed."

# 24. Visualization Engine

Creating charts for trends, comparisons, rankings and growth analysis.

In [ ]:
import plotly.express as px

def create_chart(result):

    if result["type"] == "trend":

        fig = px.line(
            result["data"],
            x="Year",
            y="Vehicles",
            color="State",
            title="Vehicle Registration Trend",
            markers=True
        )

        return fig

    elif result["type"] == "compare":

        if result.get("over_years", False):

            fig = px.line(
                result["data"],
                x="Year",
                y="Vehicles",
                color="State",
                title="State Comparison Over Years",
                markers=True
            )

        else:

            fig = px.bar(
                result["data"],
                x="State",
                y="Vehicles",
                color="State",
                title=f"State Comparison ({result['year']})"
            )

        return fig

    elif result["type"] == "compare_all":

        fig = px.bar(
            result["data"],
            x="Vehicles",
            y="State",
            orientation="h",
            title=f"Comparison ({result['year']})"
        )

        target_state = result["state"]

        fig.update_traces(
            marker_color=[
                "crimson" if s == target_state else "blue"
                for s in result["data"]["State"]
            ]
        )

        return fig

    elif result["type"] == "table":

        fig = px.bar(
            result["data"],
            x="State",
            y="Vehicles",
            title="Vehicle Registrations"
        )

        return fig

    elif result["type"] == "growth":

        fig = px.bar(
            result["data"],
            x="State",
            y="Growth",
            title="Growth in Vehicle Registrations (2006-2016)"
        )

        return fig

    elif result["type"] == "state_year":

        fig = px.bar(
            result["data"],
            x="State",
            y="Vehicles",
            title="Registered Vehicles"
        )

        return fig

    elif result["type"] == "number":

        fig = px.bar(
            x=["Total"],
            y=[result["value"]],
            title="Total Registered Vehicles"
        )

        return fig

    return None

# 25. Evaluation Framework

The system is evaluated using a representative set of benchmark questions designed to test all major capabilities supported by the analytics pipeline.

### Evaluation Objectives

The evaluation focuses on verifying:

* Ranking and sorting operations
* Trend analysis across years
* State-to-state comparison
* Aggregate computations
* Out-of-scope question detection

### Evaluation Methodology

For each benchmark question:

1. Gemini converts the natural language query into a structured query plan.
2. The query plan is executed locally using Pandas.
3. The computed result is transformed into a human-readable answer.
4. The execution status and output are recorded.

This evaluation framework ensures that both the query planning layer and the execution layer are tested across diverse analytical tasks while validating the system's ability to reject unsupported queries.


# 26. Evaluation Dataset

The following benchmark questions are used to evaluate the system's ability to perform ranking, filtering, aggregation, trend analysis, comparison, growth analysis, and out-of-scope detection.

In [ ]:
evaluation_questions = [

    # Ranking
    "Which state had the highest registered vehicles in 2016?",

    # Trend
    "Show trend of Bihar.",

    # Compare
    "Compare Maharashtra and Gujarat.",

    # Aggregate
    "What was the total number of registered vehicles in India in 2012?",

    # Out of Scope
    "What was Delhi AQI in 2016?"
]

# 27. Evaluation Results

Executing benchmark questions and recording the generated query plans, answers, and execution status.

In [ ]:
import pandas as pd
import time
import json

results = []

for q in evaluation_questions:

    try:

        plan = get_query_plan(q)

        plan = normalize_plan(plan)

        result = execute_query(plan)

        answer = generate_answer_local(q, result)

        results.append({
            "Question": q,
            "Plan": json.dumps(plan),
            "Answer": answer,
            "Status": "PASS"
        })

        # time.sleep(10)

    except Exception as e:

        results.append({
            "Question": q,
            "Plan": json.dumps(plan) if 'plan' in locals() else "",
            "Answer": str(e),
            "Status": "FAIL"
        })

evaluation_df = pd.DataFrame(results)

evaluation_df

🤖 Calling Gemini...
Calling Gemini...
Gemini responded
Response text:
```json
{
"answerable": true,
"operation": "rank",
"year": 2016,
"order": "desc",
"limit": 1
}
```
🤖 Calling Gemini...
Calling Gemini...


ERROR:tornado.access:503 POST /v1beta/models/gemini-2.5-flash:generateContent?%24alt=json%3Benum-encoding%3Dint (::1) 1557.01ms


Gemini responded
Response text:
```json
{
  "answerable": true,
  "operation": "trend",
  "state": "Bihar"
}
```
🤖 Calling Gemini...
Calling Gemini...
Gemini responded
Response text:
```json
{
"answerable": true,
"operation": "compare",
"states": ["Maharashtra", "Gujarat"]
}
```
🤖 Calling Gemini...
Calling Gemini...
Gemini responded
Response text:
```json
{
  "answerable": true,
  "operation": "aggregate",
  "year": 2012,
  "metric": "Vehicles"
}
```
🤖 Calling Gemini...
Calling Gemini...


⚠️ Gemini failed: 429 POST https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?%24alt=json%3Benum-encoding%3Dint: You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. 
* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 5, model: gemini-2.5-flash
Please retry in 2.984847035s.
🔄 Gemini disabled. Switching to Groq...
🚀 Calling Groq...


,Question,Plan,Answer,Status
0,Which state had the highest registered vehicle...,"{""answerable"": true, ""operation"": ""rank"", ""yea...",Maharashtra had 27869.87 registered vehicles.,PASS
1,Show trend of Bihar.,"{""answerable"": true, ""operation"": ""trend"", ""st...",Bihar increased from 1432.34 vehicles in 2006 ...,PASS
2,Compare Maharashtra and Gujarat.,"{""answerable"": true, ""operation"": ""compare"", ""...",Vehicle registrations in 2016:\nMaharashtra : ...,PASS
3,What was the total number of registered vehicl...,"{""answerable"": true, ""operation"": ""aggregate"",...",Total registered vehicles: 159490.00,PASS
4,What was Delhi AQI in 2016?,"{""answerable"": false, ""reason"": ""Dataset conta...",Dataset contains vehicle registration data only.,PASS


# 28. Manual Verification Query

Computing the answer directly from the dataset to validate the correctness of the generated response.

In [ ]:
manual_check = (
    df_long[
        df_long["Year"] == 2016
    ]
    .sort_values(
        "Vehicles",
        ascending=False
    )
    .head(1)
)

manual_check

,Category,State,Year,Vehicles
414,State,Maharashtra,2016,27869.866


#29. Manual Validation and Trustworthiness Check

## Objective

To verify that the system produces answers from actual dataset computation rather than relying on the language model's internal knowledge.

## Validation Question

**Question:**
Which state had the highest number of registered motor vehicles in 2016?

## System Workflow

1. The user question is interpreted by Gemini and converted into a structured query plan.
2. The query plan is executed on the motor vehicle dataset using Pandas.
3. The answer is generated exclusively from the computed result.
4. No numerical values are generated directly by the language model.

## Independent Verification

To validate correctness, the same question was answered using a direct Pandas query on the dataset:

```python
df_long[df_long["Year"] == 2016] \
    .sort_values("Vehicles", ascending=False) \
    .head(1)
```

## Results

**System Output:**

Maharashtra had the highest number of registered vehicles in 2016 with 27869.866 registered vehicles.

**Direct Dataset Query Output:**

 A direct Pandas query on the transformed dataset returned:

* State: Maharashtra
* Year: 2016
* Vehicles: 27869.866

## Conclusion

The independently computed dataset result matched the system output exactly.

This validation demonstrates that:

* Answers are derived from dataset computation rather than model memory.
* Numerical outputs are reproducible and auditable.
* The query-planning layer and execution layer are functioning correctly.
* The system provides traceable and trustworthy responses suitable for data-analysis use cases.

This approach reduces hallucination risk by separating natural-language understanding from numerical computation. The language model is used only for intent extraction, while all analytical operations are executed deterministically on the underlying dataset.


#30. System Design Decisions

## Gemini Usage Strategy

Gemini is used only for intent extraction and query planning.

All filtering, aggregation, ranking, comparison, trend analysis and growth computations are executed locally using Pandas. This minimizes API usage, improves reproducibility and ensures that numerical results originate from dataset computation rather than model memory.

## Provenance and Traceability

Every answer follows the pipeline:

User Question → Query Plan → Pandas Execution → Answer Generation

This provides transparency, auditability and easier debugging.

## Hallucination Prevention

The language model never generates numerical answers directly. Instead, it produces a structured query plan that is executed on the dataset. This significantly reduces hallucination risk.

## Out-of-Scope Handling

Questions unrelated to vehicle registration data are rejected with a clear explanation rather than generating unsupported answers.


#31. Interactive User Interface

Launching a Gradio-based interface for natural language querying of the vehicle registration dataset.

In [ ]:
import gradio as gr
import re

# -----------------------------
# Fast Local Planner
# -----------------------------
def local_query_plan(question):

    q = question.lower().strip()

    year_match = re.search(r"(20\d\d)", q)

    if year_match:

        year = int(year_match.group(1))

        if year < 2006 or year > 2016:

            return {
                "answerable": False,
                "reason": "Dataset contains vehicle registration data only for years 2006 to 2016."
            }



      # Telangana validation

    if "telangana" in q and year_match:

        if int(year_match.group(1)) < 2014:

            return {
                "answerable": False,
                "reason": (
                    "Telangana was formed in 2014. "
                    "No Telangana vehicle registration data exists before 2014."
                )
            }


    out_keywords = [
        "aqi",
        "gdp",
        "wheat",
        "rainfall",
        "temperature",
        "population",
        "accident"
    ]

    if any(word in q for word in out_keywords):

        return {
            "answerable": False,
            "reason": "Dataset contains vehicle registration data only."
        }


    # Trend

    if "trend" in q:

        states = []

        q_clean = (
            q.lower()
            .replace("&", "and")
            .replace(",", "")
        )

        for state in df_long["State"].unique():

            state_clean = (
                state.lower()
                .replace("&", "and")
                .replace(",", "")
            )

            if state_clean in q_clean:

                states.append(state)

        years = re.findall(r"(20\d\d)", q)

        if len(years) == 1:

            return {
                "answerable": False,
                "reason": (
                    "Trend analysis requires a year range. "
                    "Please specify both start and end years, "
                    "for example: Show trend of Bihar from 2014 to 2016."
                )
            }

        start_year = 2006
        end_year = 2016

        if len(years) >= 2:

            start_year = int(years[0])
            end_year = int(years[1])

        if len(states) >= 1:

            return {
                "answerable": True,
                "operation": "trend",
                "states": states,
                "start_year": start_year,
                "end_year": end_year
            }


    # Compare



    if "compare" in q:

        states = []

        q_clean = (
            q.lower()
            .replace("&", "and")
            .replace(",", "")
        )

        for state in df_long["State"].unique():

            state_clean = (
                state.lower()
                .replace("&", "and")
                .replace(",", "")
          )

            if state_clean in q_clean:

                states.append(state)

        if len(states) >= 2:

            year = 2016

            if year_match:
                year = int(year_match.group(1))


            over_years = (
                "over years" in q
                or
                "across years" in q
                or
                "throughout years" in q
            )

            return {
                "answerable": True,
                "operation": "compare",
                "states": states,
                "year": year,
                "over_years": over_years
            }



        if len(states) == 1:

            top_match = re.search(
                r"top\s+(\d+)",
                q
            )

            if (
                "other states" in q
                or
                "all states" in q
                or
                "other state" in q
                or
                "all state" in q
                or
                top_match
            ):

                year = 2016

                if year_match:
                    year = int(
                        year_match.group(1)
                    )

                limit = None

                if top_match:
                    limit = int(
                        top_match.group(1)
                    )

                return {
                    "answerable": True,
                    "operation": "compare_all",
                    "state": states[0],
                    "year": year,
                    "limit": limit
                }


    # Growth
    if any(word in q for word in [
        "growth",
        "increase",
        "expansion",
        "rise",
        "grew",
        "growth rate",
        "highest growth",
        "largest increase",
        "fastest growth",
        "most growth"
    ]):

        return {
            "answerable": True,
            "operation": "growth"
        }

    # Aggregate
    if "total" in q:

        year = 2016

        if year_match:
            year = int(year_match.group(1))

        return {
            "answerable": True,
            "operation": "aggregate",
            "year": year
        }



    # Ranking


    year = 2016

    if year_match:
        year = int(year_match.group(1))

    limit = 1

    top_match = re.search(r"top\s+(\d+)", q)

    if top_match:
        limit = int(top_match.group(1))

    if "highest" in q or "top" in q:

        return {
            "answerable": True,
            "operation": "rank",
            "year": year,
            "order": "desc",
            "limit": limit
        }

    if "lowest" in q:

        return {
            "answerable": True,
            "operation": "rank",
            "year": year,
            "order": "asc",
            "limit": limit
        }



    ranking_words = [
    "which",
    "best",
    "highest",
    "top",
    "strongest",
    "performer",
    "rank",
    "among",
    "dominated",
    "leader"
    ]

    if any(word in q for word in ranking_words):
        return None

    # State + Year Lookup


    q_clean = (
        q.lower()
        .replace("&", "and")
        .replace(",", "")
    )

    for state in df_long["State"].unique():

        state_clean = (
            state.lower()
            .replace("&", "and")
            .replace(",", "")
        )

        if state_clean in q_clean:

            year = 2016

            if year_match:
                year = int(year_match.group(1))

            return {
                "answerable": True,
                "operation": "state_year",
                "state": state,
                "year": year
            }

    return None




# -----------------------------
# Gemini Cache
# -----------------------------
query_cache = {}

def get_cached_plan(question):

    if question in query_cache:

        return query_cache[question]

    plan = get_query_plan(question)

    query_cache[question] = plan

    return plan



# -----------------------------
# Main Function
# -----------------------------





def ask_question(question):

    if not question.strip():

        return (
            "Please enter a question.",
            None
        )

    try:

        plan = local_query_plan(question)

        if plan is None:

            q = question.lower()

            dataset_keywords = [
                "vehicle",
                "vehicles",
                "registration",
                "registered",
                "state",
                "states",
                "year",
                "trend",
                "compare",
                "growth",
                "increase",
                "highest",
                "lowest",
                "top",
                "total"
            ]

            contains_state = any(
                state.lower() in q
                for state in df_long["State"].unique()
            )

            if (
                not any(
                    word in q
                    for word in dataset_keywords
                )
                and
                not contains_state
            ):

                return (
                    "Out of scope. This system only answers questions about registered motor vehicles in India (2006–2016).",
                    None
                )

            plan = get_cached_plan(question)

        plan = normalize_plan(plan)
        print("QUESTION:", question)
        print("PLAN:", plan)

        result = execute_query(plan)

        answer = generate_answer_local(
            question,
            result
        )

        fig = create_chart(result)

        return answer, fig

    except Exception as e:

        return (
            f"Error: {str(e)}",
            None
        )


# -----------------------------
# UI
# -----------------------------
with gr.Blocks(
    title="Vehicle Registration Analytics Assistant"
) as demo:

    gr.Markdown(
        """
# 🚗 Vehicle Registration Analytics Assistant

Ask natural language questions about State/UT-wise Registered Motor Vehicles in India (2006–2016).

### Supported Analytics
- Ranking
- Top-N Analysis
- Trend Analysis
- State Comparison
- Aggregation
- Growth Analysis
- Out-of-Scope Detection
"""
    )

    question_box = gr.Textbox(
        label="Ask a Question",
        placeholder="Example: Show trend of Bihar",
        lines=3
    )

    submit_btn = gr.Button(
        "Generate Answer",
        variant="primary"
    )

    answer_box = gr.Textbox(
        label="Answer",
        lines=12,
        max_lines=20
    )

    chart_output = gr.Plot(
        label="Visualization"
    )

    gr.Examples(
        examples=[
            ["Which state had the highest registered vehicles in 2016?"],
            ["Top 5 states by registered vehicles in 2016."],
            ["Show trend of Bihar."],
            ["Compare Maharashtra and Gujarat."],
            ["What was the total number of registered vehicles in India in 2012?"],
            ["Which state experienced the highest growth in vehicle registrations between 2006 and 2016?"],
            ["What was Delhi AQI in 2016?"]
        ],
        inputs=question_box
    )

    submit_btn.click(
        fn=ask_question,
        inputs=question_box,
        outputs=[
            answer_box,
            chart_output
        ]
    )

demo.launch(
    share=True,
    debug=False
)

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://284ebedb5b62ae0e79.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)
